### PR TITLE
[REVIEW] Support unary and binary ops on Indexes

### LINKED ISSUE #556: [FEA] Support comparison operations on Indexes
It would be nice for cudf GenericIndex objects to support comparison operations.  Ideally the following would work:

```python
gdf[gdf.index > 5]
```

I've started adding a test suite here: https://github.com/rapidsai/cudf/pull/555

### PR DESCRIPTION
**Pull request by kkraus14**
Fixes #556

### CODE REVIEW

**Comment by kkraus14** [comment]
These classes really need a refactor at this point, but would strongly prefer to hold off until 0.7.

File: python/cudf/dataframe/index.py [hunk 1]
```diff
@@ -117,16 +118,151 @@ def _concat(cls, objs):
...
-    def __eq__(self, other):
-        if not isinstance(other, Index):
+    def _binaryop(self, other, fn):
+        """
+        Internal util to call a binary operator *fn* on operands *self*
+        and *other*.  Return the output Series.  The output dtype is
+        determined by the input operands.
+        """
+        from cudf import DataFrame, Series
+        idx_series = Series(self)
```
Comment (thomcom) [inline python/cudf/dataframe/index.py]: I feel like you should just call `get_attr(idx_series, fn)(other)` after this line for `_binaryop`, `_rbinaryop`, and `_unaryop`. `Series` already has all these functions.

File: python/cudf/dataframe/index.py [hunk 2]
```diff
@@ -117,16 +118,151 @@ def _concat(cls, objs):
...
+
+    def __rfloordiv__(self, other):
+        return self._rbinaryop(other, 'floordiv')
+
+    def __truediv__(self, other):
+        from cudf import Series
+        from cudf.dataframe.series import truediv_int_dtype_corrections
+        if self.dtype in list(truediv_int_dtype_corrections.keys()):
+            truediv_type = truediv_int_dtype_corrections[str(self.dtype)]
+            idx_series = Series(self)
```
Comment (thomcom) [inline python/cudf/dataframe/index.py]: Again, why can't you just make it a `Series` then call `Series` binaryop?

File: python/cudf/dataframe/index.py [hunk 3]
```diff
@@ -117,16 +118,151 @@ def _concat(cls, objs):
...
+        outcol = idx_series._column.unordered_compare(cmpops, other._column)
+        result = idx_series._copy_construct(data=outcol)
+        result.name = None
+        nvtx_range_pop()
+        return as_index(result)
+
+    def _ordered_compare(self, other, cmpops):
+        nvtx_range_push("PYGDF_ORDERED_COMP", "orange")
+        from cudf import Series
+        idx_series = Series(self)
```
Comment (thomcom) [inline python/cudf/dataframe/index.py]: I think one more time to suggest the same thing.

File: python/cudf/dataframe/index.py [hunk 4]
```diff
@@ -204,18 +340,21 @@ def __getitem__(self, index):
...
 
     def __eq__(self, other):
+        return super(RangeIndex, self).__eq__(other)
+
+    def equals(self, other):
         if isinstance(other, RangeIndex):
             return (self._start == other._start and self._stop == other._stop)
         else:
-            return super(RangeIndex, self).__eq__(other)
+            return (self == other)._values.all()
```
Comment (thomcom) [inline python/cudf/dataframe/index.py]: I don't understand why you call `Index.__eq__` in one instance and create this `equals` method with other complicated logic. Shouldn't they do the same thing?

Comment (kkraus14) [reply python/cudf/dataframe/index.py]: `Index.__eq__` does an elementwise equals, so the return will be an Index of booleans. `equals` uses the elementwise equals and then gets the minimum value to determine a scalar Boolean type to return.

Comment (thomcom) [reply python/cudf/dataframe/index.py]: Oh good distinction. 👍

File: python/cudf/dataframe/index.py (hunk 1 above)
Comment (kkraus14) [reply python/cudf/dataframe/index.py]: Going to refactor to just have one helper function to handle converting to series, calling the binary op, then converting back to index.

File: python/cudf/dataframe/numerical.py [hunk 5]
```diff
@@ -104,6 +104,12 @@ def unary_operator(self, unaryop):
 
     def unordered_compare(self, cmpop, rhs):
         lhs, rhs = numeric_normalize_types(self, rhs)
+        if cmpop == 'eq':
+            if len(self) != len(rhs):
+                return False
+        elif cmpop == 'ne':
+            if len(self) != len(rhs):
+                return True
```
Comment (thomcom) [inline python/cudf/dataframe/numerical.py]: Why does this go into unordered compare, but not into ordered compare? It seems like it should go into `numeric_column_compare`?

Comment (kkraus14) [reply python/cudf/dataframe/numerical.py]: This is actually wrong period, this should be for `.equals` and not an elementwise operation. Thanks!

**Pull request merged**